### ISSUE TITLE
mass file formatting

### ISSUE THREAD

**Issue by DanVanAtta** [triplea-game/triplea#19]
- I can help do mass file formatting, I'm pretty good with the bash-fu (help is welcome though)
- If we do mass formatting, is their bandwidth to review it? If so, what is testing criteria? Can we skip testing so long as all format actions are done in automated (repeatable ways)?
- If that sounds good, what things should we  format?

To start the list of things to format (please comment on additional items to be mass formatted):
- remove license, `@author`, `@version`, empty javadoc
- switch to new format style, if we switch (#18)
- add curly braces to all if statements

**Comment by ron-murhammer** [comment]
I think we accumulate this list then try to do them sort of all at once preferably right before or after a release. This will minimize the risk of conflicting with ongoing development. In regards to your initial list:

1. Agree with most of these but should probably think about general commenting standards
2. Agree. Should discuss formatting and standards.
3. This is part of #2 IMO.

**Comment by veqryn** [comment]
@ron-murhammer @DanVanAtta i posted earlier that i added the google format xml to our project with minor changes.
I ran eclipse cleanup using that, to add override tags and format all code.  Please see pull #23 
Let me know what you think.

**Comment by ron-murhammer** [comment]
Looks good and agree with the new format (honestly hated the old one). Just merged it.

**Comment by DanVanAtta** [comment]
:+1: 
Down the road if we need to upgrade the formatter, noting the difference we applied will make upgrade easier. Alternatively when that day happens we'll lose the tweaks or will just need to remember them (probably looking at diff of new format vs old will indicate what it is too)

(Generally a wiki with formatting decisions likely will be helpful.)

@veqryn Do you plan to add in @Override annotation? Eclipse can add it to everywhere in one go (though that required me to increase eclipse heap size. In eclipse preferences you can toggle warnings/error/ignore, missing @Override can be changed from ignore to warning. Then in the problems list, you can select one of the @Override missing warnings, should have a box of all the instances of the problem where you can select it all).

Here is the command to remove @author from javadoc comments:
for i in $(find . -name "*.java"); do sed -i '/@author/d' $i; done
(if you are using a mac the sed command would actually be IIRC: sed -i '' -e '/@author/d')

Here is the xargs version of the same:
find . -name "*.java" | xargs sed -i '/@author/d'

**Comment by DanVanAtta** [comment]
Started this wiki page to document some of the format decisions from this issue (bit of a shell of a page right now): https://github.com/triplea-game/triplea/wiki/Code-format

Future changes can probably be done via Pull Requests, the wiki is a good place to document the format choices and to do so in the future. I think this issue is good to close, anything further to discuss?

**Comment by DanVanAtta** [comment]
You can also get rid of the empty @return in javadocs with this:
$ find . -name "*.java" | xargs sed -i '/@return$/d' 
or
$ for i in $(find . -name "*.java"); do  sed -i '/@return$/d' $i; done


The  command below can help verify that it is safe to do. The -A1 returns the next line after the @return match, then we uniq the ouptu and see we have 4 cases. I checked the last 3 by hand they are okay, the @return is empty. On the other hand, we do (did have @param) annotations where the comment is on the next line. Not sure yet where the google format will place the comment for @param.

$ find . -name "*.java" | xargs grep -h -A1 "@return$" | grep -v "return" | sort | uniq
--
	 */
	public int getUnitImageHeight(final Image img)
	public int getUnitImageWidth(final Image img)
	 * @throws GameParseException

**Comment by veqryn** [comment]
I will take a look and try this out.  I am a little possessive and would like all "mass" file changes to go through me for a bit.

**Comment by DanVanAtta** [comment]
Closing..

dos2unix changes, new code formatting and @Override were big changes and have been merged : )

Only remaining items that I can think of would be the @author and empty @return that can easily be removed. @veqryn I can understand, hence why I gave you the commands rather than submitting pull requests ;)

**Comment by veqryn** [comment]
@DanVanAtta I am trying to delete empty comment blocks. This works in my regex programs, but does not work with SED for some reason. Can you tell me why, and if possible fix it?
find . -type f -name "*.java" | xargs sed -i '/[ ]*\/[*]+[\s*]*[*]+\//d'

**Comment by veqryn** [comment]
Also how to get rid of everything before the first instance of the word 'package' that isn't inside a comment?
find . -type f -name "*.java" | xargs sed -i 's/^[\s\S]+package/package/'
does not seem to work...

**Comment by gaborbernat** [comment]
@veqryn 

this does work for me for empty comment lines:
 find -type f -name *.java  | xargs -I {} sed -r -e '/^[ \t]*\/\/[ \t]*$/d'  -i {}

**Comment by gaborbernat** [comment]
@veqryn 
can't understand what you are trying to do with the second, perhaps can you provide an example?

**Comment by DanVanAtta** [comment]
I was using this to find the empty comments, not too many of them, bernat beat me to it.
find . -name "*.java" | xargs grep -P "// *\t*$" | sort | uniq

**Comment by gaborbernat** [comment]
side note and fun fact, went with xargs for compatibility, but for anyone working with shell scripts and xargs should definitely look into parallel, beats xargs hands down, and way more powerful, https://www.gnu.org/software/parallel/parallel_tutorial.html#GNU-Parallel-Tutorial

**Comment by veqryn** [comment]
Ok, so github is removing stuff that we type, which is why some of this stuff isn't working. Half my special characters got removed.  We need to use code markdown.
Quote your code in single or triple backticks please.

**Comment by gaborbernat** [comment]
done :)

**Comment by veqryn** [comment]
I am trying to remove ALL the comments before the package name line, and I am also trying to remove all empty comment blocks (they were made empty when I earlier removed empty `@returns` and `@version` and `@author` etc).

```
/*
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 */
/*
 * IntegerMapTest.java
 *
 * Created on November 7, 2001, 1:46 PM
 */
package games.strategy.util;

import junit.framework.TestCase;

/**
 *
 */
public class VersionTest extends TestCase {
  /** Creates new IntegerMapTest */
  public VersionTest(final String name) {
    super(name);
  }
```

**Comment by DanVanAtta** [comment]
For a first step for the file headers, you can do this:

`for i in $(find . -name "*.java"); do head -$(grep -hn "^package " $i | sed 's/:.*//g') $i; done`

That shows you everything above the line where the word "package" starts the line. From here you can see everything above it is garbage.

To capture it in a file, do this:
`for i in $(find . -name "*.java"); do head -$(grep -hn "^package " $i | sed 's/:.*//g') $i; done >> outputFileName`   though, it's not a perfect one line, so after about 30 seconds, grep will be waiting for input from standard in, so hit ctrl+c to end it, then you can do this;  `sort outputFileName | uniq` to inspect it better.

I think this type of strategy probably will be used to chop off the head of the files. I was hoping awk could do it, but looks it cannot either. sed will not be able to do it since it works line by line.

I think I can think of some clever ways to get rid of the empty line comments.

**Comment by gaborbernat** [comment]
for multiline matching sed is really troublesome to handle, would probably just look towards perl

**Comment by gaborbernat** [comment]
this should do the job:
```find -type f -name *.java  | xargs -I {} perl -i -pe 'BEGIN{undef $/;} s/.*?package/package/smg' {}```

**Comment by DanVanAtta** [comment]
Doing this in bash is not too bad, but you need to filter to only files that need the operation. Basically you can tail starting from a line matching "^package" and then redirect that tail to the same file. 

@gaborbernat @veqryn  has the answer for both questions though.

I ran the one line to view the file tops, it's empty now after having ran the perl command suggested:
`find . -name ".java" | xargs grep -P "// *\t$" | sort | uniq`

@gaborbernat  is right, for multi-line perl is probably better. For bash you need to do hacks. Like replacing all new lines with a unique character, operating, then replacing the unique character again, other stuff like that.

**Comment by veqryn** [comment]
Yep, perl was a ton easier than trying to hack bash and sed.
Here is what I went with for this next branch pull:
```
find . -type f -name "*.java" | xargs sed -i '/@author/d'
find . -type f -name "*.java" | xargs sed -i '/@version/d'
find . -name "*.java" | xargs sed -i '/@return$/d'
find . -name "*.java" | xargs sed -i '/@param$/d'
find . -name "*.java" | xargs sed -i '/^[ \t]*\/\/[ \t]*$/d'
find . -type f -name "*.java" -exec perl -0777 -pi -e 's/^[\s\S]+package game/package game/' {} \;
find . -type f -name "*.java" -exec perl -0777 -pi -e 's/^[\s\S]+package util/package util/' {} \;
find . -type f -name "*.java" -exec perl -0777 -pi -e 's/[ ]*\/[*]+[\s*]*[*]+\///' {} \;
```

**Comment by DanVanAtta** [comment]
Two more things that can be mass-formatted away:

These comments now that we have @override:
  /*
   * (non-Javadoc)
   *
   * @see games.strategy.engine.framework.startup.launcher.ILauncher#launch(java.awt.Component)
   */


Second, empty @param in javadoc. I suspect once you re-wrap, the @param comment will always start on the same line. In that case the pattern  "@param (word) (end of line)" should match the empty javadoc comments.

**Comment by ron-murhammer** [comment]
So I looked over the latest code and I'm not particularly thrilled by the changes. It seems a lot of lines have been wrapped and look awful and difficult to read now. I think I actually preferred it before all the changes...

**Comment by DanVanAtta** [comment]
Re-wrap likely would help, that caused most of the problems from what I saw. Re-wrap is not good on an ongoing basis since it manually wipe out your own manual line wrappings.

I'll mention too formatting only goes so far. Lipstick on a pig and all...

**Comment by ron-murhammer** [comment]
Agree. I think we need to re-wrap as there are just tons of lines that were slightly too long that got wrapped with like 10 characters on a line then another wrapped line that need combined. I think it also has to do with the bracket change which I prefer in general but had coded with very few spaces since brackets on separate lines added so much. Need to go back and add spaces to make things more readable.

**Comment by DanVanAtta** [comment]
Add spaces? Like white lines?

**Comment by ron-murhammer** [comment]
Yeah. When you have 20-30 lines with no white lines breaking it up its tough to read. Where before you had brackets on separate lines that tended to break it up more.

**Comment by DanVanAtta** [comment]
Personally I prefer to extract methods. 20-30 line method is probably doing too much. 4 or 5 high level functions keeps it easy to understand, and 4 or 5 liner functions are simple small pieces which each can be easier to deal with on their own.

But yeah, sometimes the old new line is simply needed..

**Comment by DanVanAtta** [comment]
Let's close this thread. I urge us not to do a lot more mass reformatting while we have so many in-flight PR's now that touch a lot of files. Be nice to get everything integrated and avoid merge conflicts.

Re-open if there really is more discuss.

We've discussed to death line lengths, 120vs130vs140. The conclusion being that @veqryn will possibly reformat the code with re-wrap and an updated line length.

**Comment by veqryn** [comment]
I'll do a re-wrap today, and also kill annotations: see, since, and (non-Javadoc).
After that, we can take a break from mass formatting for a very very long time.

Some things to note:

The mass cleanup erased any comments that were on the same line as a { or the if/while/else/function/class/etc above it.
So please do not put comments in spots that have a brace or might have a brace.

Secondly, the mass formatting did some weird things to comments that were on the same line as code, but went beyond 140 characters.  It wrapped them down, but then had them indented to where-ever the original comment started.  This sometimes resulted in comments being wrapped with 1 letter or 1 word per line, going 10 lines down.  It is ugly, but did not happen very often, so I do not plan on reverting the merge.  Please fix them if you see them.

**Comment by DanVanAtta** [comment]
Good, some useful stuff there. I'm happy you went back and did the re-wrap.

I'm not completely happy:
- we had data that showed 130 line length was just as good as 140, that was ignored along with other good reasons for not going with such long lines. I think in practice we'll find that lines as long as 140 characters long will be manually re-wrapped to be shorter anyways. That diminishes the value of automatic formatting. Though, a saving grace is that with "re-wrap" off we'll avoid formatting wars - so it's not so bad.

- there were a number of merge conflicts, I felt resolving them was needless and a couple hours that could have been better spent elsewhere. Most of the PRs open are easy to review and can be quickly closed within a minute (I'm thinking of this one as an example: #33 , a two file change with a total of two lines added to javadoc two classes. Opening the PR probably will take longer than reviewing it.)

. Let me put it this way: we just reformatted dead code that was removed in another branch, resulting in merge conflicts. The dead code had to removed yet once again and then re-pushed. IMO that is pretty clearly not the most effective way to collaborate.

**Comment by DanVanAtta** [comment]
It's *really* nice we got rid of the old formatting and cleaned up so much of the commentary. Make the code just that more reasonable.

Closing this issue. Future work can probably be raised as a  PR, discussed as a new issue, or re-open this one.

**Comment by veqryn** [comment]
Actually you may have misread.  I said that I'd do the re-wrap today... #65 

And I did not mean to step on your toes.  I thought we agreed to do the cleanup first, before lots more pull requests...

**Comment by DanVanAtta** [comment]
Ah, I found the comment that anything beyond 120 doesn't look good in comparison tool. :+1: 
Thanks, no worries about the slight toe steppage.

**Comment by ron-murhammer** [comment]
@veqryn I just imported the latest formatting profile and noticed that it doesn't seem to match exactly with the format of the current code as I tested just clicking format on a few files and they had some minor changes. Can you make sure you have the latest formatting profile here: https://github.com/triplea-game/triplea/blob/master/triplea_java_eclipse_format_style.xml

**Comment by veqryn** [comment]
The only difference should be comments that exceed 120 characters.  Those did not get cleaned up, because they really need a human touch, to prevent the same problems that happened the first time I did a mass formatting.  So those I sort of left as a 'they will get fixed as people touch the files' sort of thing.  I am much less concerned about long comments then long code, though I do want a good format standard in place for future code.
If it is anything other than comments, please let me know, because then it is probably something I am not aware of.

**Comment by ron-murhammer** [comment]
There appear to be some other issues as well. You can see my latest commit  (https://github.com/triplea-game/triplea/commit/7f208676db5b9a8f87975e07991226d6b358a134) has formatting changes.

Things I noticed when I formatted the files I updated (most occur multiple times):
- import sorting (line 3: maybe I have an additional setting on?)
- line wrapping (line 65: seems that some lines were still over the 120 max?)
- comment wrapping (line 229)
- wrapped line indent change (line 665)

**Comment by veqryn** [comment]
Ya, i think you have additional settings on.  The import sorting should definitely not happen like that.
My guess is your are using eclipse, and your cleanup script is different than the one I exported and included...

**Comment by DanVanAtta** [comment]
Dev setup wiki page updated to include mention of the cleanup script: https://github.com/triplea-game/triplea/wiki/%23DevSetup--Code-format

Closing, doesn't seem like there is anything more to discuss at the moment.

**Comment by ron-murhammer** [comment]
So there appears to be a difference between the formatting profile and what you actually formatted. The major difference I see is that comments appear to be formatted at around 140 vs code is formatted around 120. The formatting profile specifies them both to be 120 so every time I format a file it wraps a whole bunch of comments that are at 140. I believe the profile is correct as we want 120 for comments as well and you need to reformat all the code again.

**Comment by veqryn** [comment]
@ron-murhammer: Like I said above, the comments being wrapped at 120 was not included in the "Auto" formatting PR because eclipse's cleanup and formatting of those comments is generating issues, such as breaking up a comment into 1 line per word, stretching down many lines.
This is why I said they need a human touch.

So, when doing a commit, if touching a file where the comments are getting f-ed up, then please fix them by hand.

**Comment by ron-murhammer** [comment]
In that case can we just change the formatting profile to either not wrap comments or have a greater width for them. Otherwise if I edit 1 line in a 1000 line file and then have 100 lines of comments wrap I need to go fix 100 comment lines.

**Comment by veqryn** [comment]
We are trying to fix them as we see them.  
You can see if you have done that before committing by doing 'git diff'.
If someone has time, they can run the eclipse cleanup thing, and it will show all the files where the comments are getting wrapped, and then go in an fix the ones where it is ugly.

**Comment by ron-murhammer** [comment]
I'd prefer to either just format all the comments to be 120 then fix them to look better as we go or increase the comment width so it doesn't auto format all the comments in a file that I change 1 line in.

**Comment by DanVanAtta** [comment]
Don't you have control over when format is done @ron-murhammer ? Turn off the format on save option? In general I agree with @ron-murhammer  on fixing it all so format changes are an exception and not common.

**Comment by DanVanAtta** [comment]
Next action item in this ticket: reformat comments to 120.

Are we agreed to that action item, or should we close as going to continue fix formatting ad-hoc?

**Comment by veqryn** [comment]
Would a particular someone with lots of time, hint hint, wink wink, like to reformat all comments to less than 120 length (fixing by hand all instances where the comment needs to go on the above or below line, because eclipse cleanup is f-ing it up)?
I will definitely quickly accept a patch that only touches comments (though please be careful about not deleting comments outright).

**Comment by gaborbernat** [comment]
Did you use the latest Eclipse 4.5? The formatter has been heavily rewritten in that; with lot of bug fixes. Maybe it would work just fine with the new version :-? just a  thougth. see https://waynebeaton.wordpress.com/2015/03/15/great-fixes-for-mars-winners-part-i/ for details :)

**Comment by veqryn** [comment]
I do, and it is much less buggy on a particular work project I have, but it still has this issue for comments.

**Comment by gaborbernat** [comment]
Can you please explain the issue, I would to at least report the bug :) if it's really a bug :) as I've already contacted Mateusz for supporting one line for loops :)

**Comment by veqryn** [comment]
I'm not sure it really is a bug.  It is just ugly as hell.
For an example, look at the before and after of commit d01944fc4c8b42a1754b47e06a06ba15106ecdbe, on file MoveValidator.java
You can see what it does to one of the comments, splitting the comment into 20 lines long.

**Comment by DanVanAtta** [comment]
#149  open to fix the comment formatting. 

MoveValidator.java and other examples should pretty much all be fixed. Comment formatting is not going to be perfect, though we can now safely run the formatter everywhere and avoid the end of line commentary mess.

I'm going to close this issue now in favor of the PR, which should resolve the outstanding mass format issues.

**Comment by DanVanAtta** [comment]
Since I can't attach a link to the fix for `d01944f`, posting a screenshot (as an example) showing the formatting fix for MoveValidator.java:
![movevalidatordiff](https://cloud.githubusercontent.com/assets/12397753/9722905/8cc5b16e-556a-11e5-8920-48dce6113f56.png)
